pkg/standalone-utils/test/LidoRelayer.test.ts

```diff
@@ -1,462 +1,950 @@
+    token: Token
+  ) {
+    return expectEvent.inIndirectReceipt(receipt, token.instance.interface, 'Transfer', args, token.address);
+  }
+
+  describe('primitives', () => {
+    describe('wrapStETH', () => {
+      let tokenSender: Account, tokenRecipient: Account;
+
+      context('sender = user, recipient = relayer', () => {
+        beforeEach(() => {
+          tokenSender = sender;
+          tokenRecipient = relayer;
+        });
+        testWrap();
       });
 
-      context('when the user did allow the relayer', () => {
-        sharedBeforeEach('allow relayer', async () => {
-          await vault.instance.connect(sender).setRelayerApproval(sender.address, relayer.address, true);
+      context('sender = relayer, recipient = relayer', () => {
+        beforeEach(async () => {
+          await stETH.transfer(relayer, fp(1), { from: sender });
+          tokenSender = relayer;
+          tokenRecipient = relayer;
         });
+        testWrap();
+      });
 
-        describe('swap using stETH as an input', () => {
-          let singleSwap: SingleSwap;
+      context('sender = relayer, recipient = sender', () => {
+        beforeEach(async () => {
+          await stETH.transfer(relayer, fp(1), { from: sender });
+          tokenSender = relayer;
+          tokenRecipient = sender;
+        });
+        testWrap();
+      });
 
-          sharedBeforeEach('build swap request', async () => {
-            singleSwap = {
-              kind: SwapKind.GivenIn,
-              poolId: basePoolId,
-              assetIn: basePoolTokens.findBySymbol('wstETH').address,
-              assetOut: basePoolTokens.WETH.address,
-              amount: fp(1),
-              userData: '0x',
-            };
-          });
+      function testWrap(): void {
+        it('wraps with immediate amounts', async () => {
+          const amount = fp(1);
+          const expectedWstETHAmount = await wstETH.instance.getWstETHByStETH(amount);
 
-          it('performs the given swap', async () => {
-            const receipt = await relayer.connect(sender).swap(singleSwap, funds, limit, deadline);
+          const receipt = await (
+            await relayer.connect(sender).multicall([encodeWrap(tokenSender, tokenRecipient, amount)])
+          ).wait();
 
-            expectEvent.inIndirectReceipt(await receipt.wait(), vault.instance.interface, 'Swap', {
-              poolId: singleSwap.poolId,
-              tokenIn: singleSwap.assetIn,
-              tokenOut: singleSwap.assetOut,
-              amountIn: singleSwap.amount,
-              // amountOut
-            });
-          });
+          const relayerIsRecipient = TypesConverter.toAddress(tokenRecipient) === relayer.address;
+          expectTransferEvent(
+            receipt,
+            {
+              from: relayerIsRecipient ? ZERO_ADDRESS : relayer.address,
+              to: relayerIsRecipient ? relayer.address : TypesConverter.toAddress(tokenRecipient),
```

Comment:
Note that we're not testing that the tokens came from the sender: maybe we should have two assertions, one where the sender sends the tokens, and one where the recipient receives them.

Reply:
The same appies to other `testFoo()` tests.